### PR TITLE
Add an instrumentation component that can be used to instrument certain statements and expressions during lowering.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Move sequence point injection from LocalRewriter to the instrumentation component.
Related to #9819.

Note, I haven't got rid of all places where LocalRewriter deals with sequence points yet. 

@JohnHamby, @dotnet/roslyn-compiler Please take a look.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
CC @jcouv

**Comment by JohnHamby** [comment]
The approach makes sense to me and the code seems comprehensible. I don't have enough context to know if all the details are correct.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LocalDeclaration.cs [hunk 1]
```diff
@@ -66,16 +66,18 @@ private BoundStatement RewriteLocalDeclaration(CSharpSyntaxNode syntax, LocalSym
...
                 hasErrors);
 
-            return AddLocalDeclarationSequencePointIfNecessary(syntax, localSymbol, rewrittenLocalDeclaration, wasCompilerGenerated);
+            return AddLocalDeclarationSequencePointIfNecessary(originalOpt, localSymbol, rewrittenLocalDeclaration);
         }
 
-        private BoundStatement AddLocalDeclarationSequencePointIfNecessary(CSharpSyntaxNode syntax, LocalSymbol localSymbol, BoundStatement rewrittenLocalDeclaration, bool wasCompilerGenerated = false)
+        private BoundStatement AddLocalDeclarationSequencePointIfNecessary(BoundLocalDeclaration originalOpt, LocalSymbol localSymbol, BoundStatement rewrittenLocalDeclaration)
         {
             // Add sequence points, if necessary.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LocalDeclaration.cs]: I'm assuming you have planned this for later update, but any mention of sequence point should be removed from local rewriter and replaced with mentions of instrumenter.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LocalDeclaration.cs]: I am planning to rename the method, but I plan to keep some of the comments as they give an idea why did we choose to provide instrumentation for certain things.

**Comment by jcouv** [comment]
LGTM. 
I might suggest `SequencePointInstrumenter` instead of `DebugInfoInjector`.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LocalDeclaration.cs (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LocalDeclaration.cs]: That works.

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/CompoundInstrumenter.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal class CompoundInstrumenter : Instrumenter
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/CompoundInstrumenter.cs]: Docs please.

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs [hunk 3]
```diff
@@ -0,0 +1,322 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal class DebugInfoInjector : CompoundInstrumenter
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs]: Docs please

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Diagnostics;
+using System;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal class Instrumenter
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs]: Doc, please

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 5]
```diff
@@ -68,7 +73,9 @@ internal sealed partial class LocalRewriter : BoundTreeRewriterWithStackGuard
             try
             {
                 var factory = new SyntheticBoundNodeFactory(method, statement.Syntax, compilationState, diagnostics);
-                var localRewriter = new LocalRewriter(compilation, method, methodOrdinal, containingType, factory, previousSubmissionFields, allowOmissionOfConditionalCalls, diagnostics);
+                var localRewriter = new LocalRewriter(compilation, method, methodOrdinal, containingType, factory, previousSubmissionFields, allowOmissionOfConditionalCalls, diagnostics,
+                                                     DebugInfoInjector.Singleton);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: Why are we *always* using the DebugInfoInjector?

**Comment by gafter** [comment]
LGTM, but please doc why we always use the debug injector.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: Because this is what compiler does today, it always injects sequence points.

**Comment by AlekseyTs** [comment]
@gafter I am not sure what documentation you have in mind. This reflects the current behavior of the compiler. I have to do this otherwise tests would fail.

**Comment by gafter** [comment]
i am suggesting to add comments in the code to explain why, logically, that injector is correct to use in all cases.

**Comment by AlekseyTs** [comment]
Why wouldn't it be? This is a refactoring, I am not changing any behavior. It doesn't feel useful to add a comment that says that. Could you perhaps give an example of the comment you would like to see added.

**Comment by gafter** [comment]
Something like
`// We use the DebugInfoInjector unconditionally in the local rewriter because...`
and then explain why. Referring to previous versions of the compiler isn't helpful to the reader, as the reader is generally interested in understanding the behavior of the version of the compiler being inspected.

**Comment by AlekseyTs** [comment]
I am not talking about native compiler. I am talking about this compiler before and after this change. I haven't seen any comment explaining why this compiler does this all the time. And I don't know exact reason. Unless you can provide an explanation, I have nothing to say about the reasons.

**Comment by AlekseyTs** [comment]
I don't need any special reason to preserve existing behavior.

**Comment by gafter** [comment]
"pre-existing behavior" won't mean anything to someone reading the code after this change, so that's probably not a good way to explain it. Perhaps understanding why tests break would help understand why things are this way, but it is better to do that now and add comments rather than expecting future readers to do that experiment.

**Comment by AlekseyTs** [comment]
The tests break because they expect the sequence points be injected all the time. This behavior is not an accident, I don't remember the reasons behind this and, frankly, don't believe your request is related to the change I am making. Fell free to open an issue about adding an explanation why LocalRewriter always adds sequence points. Alternatively, you can provide an explanation and I add it.

**Comment by gafter** [comment]
Are you saying there are tests that test the shape of the initial bound trees, and those tests require us to insert sequence points in non-debug builds?

**Comment by AlekseyTs** [comment]
I am saying that I am preserving existing behavior. Your questing about why existing behavior what it is might have some merits, but it is outside of the scope of my change.

**Comment by gafter** [comment]
I'm saying that the code there (where `DebugInfoInjector.Singleton` is passed directly to the `LocalRewriter` constructor) deserves a comment for the reader, as it is counterintuitive.

**Comment by AlekseyTs** [comment]
Noted

**Comment by AlekseyTs** [comment]
@JohnHamby, @jcouv, @dotnet/roslyn-compiler, @drognanar, @genlu, @tmat, @ManishJayaswal  Please review, I completed the refactoring and would like to merge it in if there are no objections.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 6]
```diff
@@ -254,6 +254,7 @@ static void TestMain()
 Flushing
 1
 True
+True
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: This looks odd. What introduces the '+' character?

**Comment by JohnHamby** [comment]
LGTM. I didn't study all the details, but I like the approach very much.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: You are correct, transformation of code coverage instrumentation to this architecture is still in the future. The instrumentation point is introduced for implicit constructor initializer of an explicitly declared constructor, we create a sequence point for it. If you don't think it should be instrumented for code coverage, I can suppress that. Let me know.

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs [hunk 7]
```diff
@@ -75,6 +75,10 @@ public int Count
             {
                 return _builder.Count;
             }
+            set
```
Comment (VSadov) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: Is this used in the same way as Clip or it is to support expansion of the builder with uninitialized values as well?

Comment (AlekseyTs) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: The change uses it as Clip, for some reason Clip didn't catch my eye. Would you prefer me to undo this change?

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This type is responsible for adding debugging sequence points for the executable code.
+    /// It can be combined with other <see cref="Instrumenter"/>s. Usually, this class should be 
+    /// the root of the chain in order to ensure sound debugging experience for the instrumented code.
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs]: I assume this means that sequence points transformation is typically applied after all other changes. I am not sure comment is clear enough about this part.

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    /// </summary>
+    internal partial class DebugInfoInjector : CompoundInstrumenter
+    {
+        /// <summary>
+        /// A singleton object that performs only one type of instrumentation - addition of debugging sequence points. 
+        /// </summary>
+        public static readonly DebugInfoInjector Singleton = new DebugInfoInjector(Instrumenter.NoOp);
+
+        public DebugInfoInjector(Instrumenter previous)
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs]: `private`?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs]: No, it is my intent to keep the constructor accessible. For example, for cases when we need to combine sequence points with code coverage instrumentation.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler I need a second sign-off from the compiler team for this change.

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        {
+            Debug.Assert(!original.WasCompilerGenerated);
+            return rewritten;
+        }
+
+        public virtual BoundStatement InstrumentNoOpStatement(BoundNoOpStatement original, BoundStatement rewritten)
+        {
+            return InstrumentStatement(original, rewritten);
+        }
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs]: double blank line

**Comment by jcouv** [comment]
LGTM.

**Comment by jaredpar** [comment]
:+1:

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: @JohnHamby Would you like me to restore the original behavior of the code coverage instrumentation?

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: @VSadov Are you Ok with keeping the set accessor?

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs (hunk 8 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs]: Adjusted the comment.

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs]: Removed extra blank line.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs (hunk 6 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: We've been trying to limit code coverage instrumentation to actual source code, so I would prefer the original behavior, but I'm certainly open to discussing this.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: @JohnHamby I submitted PR #10499 to address this.
